refactor: update cloud function deployment

Marco-Premier opened the pull request:
Cloud Functions not redeployed when config or env vars change. Expand triggers_replace to include env vars, secrets, and config file generation.

Fixes #3986

Issue: #3986

SanjayVas reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: 0 of 5 files reviewed, 1 unresolved discussion

___
*[`src/main/terraform/gcloud/modules/http-cloud-function/variables.tf` line 57 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073#-Obi5AC-E_naTSVpYGOr:-Obi5AC-E_naTSVpYGOs:b-ek6am7) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/96e6d9bbdcf4a8fe451ed402c3beb234811bfbda/src/main/terraform/gcloud/modules/http-cloud-function/variables.tf#L57)):*
> ```HCL
> }
> 
> variable "config_path" {
> ```

I don't like this taking in the config path solely for the purpose of redeploying. If the Terraform module doesn't actually need to read the config, it shouldn't take it as a variable.

Since the config is read from a storage bucket and all that's needed to re-read it is to restart (but not re-deploy) the function, we would just update the GHA workflow to unconditionally trigger restart. Restarting on every deployment is fine, but what I want to avoid is unnecessary deployment.


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: 0 of 5 files reviewed, 1 unresolved discussion (waiting on @Marco-Premier)

___
*[`src/main/terraform/gcloud/modules/http-cloud-function/variables.tf` line 57 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073#-Obi5AC-E_naTSVpYGOr:-Obi5hflEP08fIVupc2s:bt0stpx) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/96e6d9bbdcf4a8fe451ed402c3beb234811bfbda/src/main/terraform/gcloud/modules/http-cloud-function/variables.tf#L57)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I don't like this taking in the config path solely for the purpose of redeploying. If the Terraform module doesn't actually need to read the config, it shouldn't take it as a variable.

Since the config is read from a storage bucket and all that's needed to re-read it is to restart (but not re-deploy) the function, we would just update the GHA workflow to unconditionally trigger restart. Restarting on every deployment is fine, but what I want to avoid is unnecessary deployment.
</blockquote></details>

Sorry, overloaded "deployment" there. Restarting every time we're deploying to a Halo environment is fine, but I want to avoid unnecessary recreation of resources.


<!-- Sent from Reviewable.io -->

Marco-Premier reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: 0 of 5 files reviewed, 1 unresolved discussion (waiting on @SanjayVas)

___
*[`src/main/terraform/gcloud/modules/http-cloud-function/variables.tf` line 57 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073#-Obi5AC-E_naTSVpYGOr:-Oc1D2mZ5AOkMDAO_wr4:b-cox2ny) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/96e6d9bbdcf4a8fe451ed402c3beb234811bfbda/src/main/terraform/gcloud/modules/http-cloud-function/variables.tf#L57)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Sorry, overloaded "deployment" there. Restarting every time we're deploying to a Halo environment is fine, but I want to avoid unnecessary recreation of resources.
</blockquote></details>

There is the

gcloud run services update --update-env-vars ... --set-secrets
that can be used to create a new revision, meaning: restarting the cloud function.

Therefore if those changes, a new block:

provisioner "local-exec" {
...
}

can run to restart the function without re-deploying.
However with this approach, if (only)  the content of the function configs change, it becomes not possible to restart the function.

To conclude, I may use 

1. gcloud run services update command, to restart the function
2. redeploy the function if either the image or the config content changes

I like more what I have already in this PR, as it really unlikely that env vars and secrets would change without requiring also the config content to change


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (changes requested):
> @SanjayVas reviewed 4 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: 4 of 5 files reviewed, 1 unresolved discussion (waiting on @Marco-Premier)

___
*[`src/main/terraform/gcloud/modules/http-cloud-function/variables.tf` line 57 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073#-Obi5AC-E_naTSVpYGOr:-Oc1zacU4Xlcifq12FnK:bir05yu) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/96e6d9bbdcf4a8fe451ed402c3beb234811bfbda/src/main/terraform/gcloud/modules/http-cloud-function/variables.tf#L57)):*
> What I'm suggesting is to not use Terraform at all just to restart the functions when they don't need to be redeployed. Instead, trigger the restart in the GHA workflow

So it sounds like there's no way to trigger picking up config content changes from the bucket without redeploying.

There are two potential approaches here:

1. Switch to an intentionally coupled solution where config content changes are tied to deployment.
2. Figure out how to decouple these and be able to re-read configs without redeployment.

Assuming we want to do (1) and still stay with the storage bucket solution, then what we need is to tie writing the bucket from the GitHub environment to deploying the function. This PR is currently trying to condition based on changes to the GitHub environment as opposed to changes to the files in the bucket. 

Can we trigger based on the resources that do the upload to the bucket?


<!-- Sent from Reviewable.io -->

Marco-Premier reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: 4 of 5 files reviewed, 1 unresolved discussion (waiting on @SanjayVas)

___
*[`src/main/terraform/gcloud/modules/http-cloud-function/variables.tf` line 57 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073#-Obi5AC-E_naTSVpYGOr:-OcBkRZF6ve0MwRNmGgC:bx5e7pj) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/96e6d9bbdcf4a8fe451ed402c3beb234811bfbda/src/main/terraform/gcloud/modules/http-cloud-function/variables.tf#L57)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> What I'm suggesting is to not use Terraform at all just to restart the functions when they don't need to be redeployed. Instead, trigger the restart in the GHA workflow

So it sounds like there's no way to trigger picking up config content changes from the bucket without redeploying.

There are two potential approaches here:

1. Switch to an intentionally coupled solution where config content changes are tied to deployment.
2. Figure out how to decouple these and be able to re-read configs without redeployment.

Assuming we want to do (1) and still stay with the storage bucket solution, then what we need is to tie writing the bucket from the GitHub environment to deploying the function. This PR is currently trying to condition based on changes to the GitHub environment as opposed to changes to the files in the bucket. 

Can we trigger based on the resources that do the upload to the bucket?
</blockquote></details>

For option2:

at the moment terraform uploads the config file at every run. I couldn't find any trigger on the documentation that supports the triggering based on the file content:
https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/storage_bucket_object

which means that even if the cloud function depends on the cloud storage upload, that happens at every run. So I don't think it's doable to link the re-deployment of the function based on whether the file is uploaded or not.


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: 4 of 5 files reviewed, 1 unresolved discussion (waiting on @Marco-Premier)

___
*[`src/main/terraform/gcloud/modules/http-cloud-function/variables.tf` line 57 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073#-Obi5AC-E_naTSVpYGOr:-OcCd6Ok5gys_hKpRjoL:b-8mznpi) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/96e6d9bbdcf4a8fe451ed402c3beb234811bfbda/src/main/terraform/gcloud/modules/http-cloud-function/variables.tf#L57)):*
> I couldn't find any trigger on the documentation that supports the triggering based on the file content

Doesn't `source_md5hash`[^1] do this?

[^1]: https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/storage_bucket_object#source_md5hash-1


<!-- Sent from Reviewable.io -->

Marco-Premier reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: 1 of 8 files reviewed, 1 unresolved discussion (waiting on @SanjayVas)

___
*[`src/main/terraform/gcloud/modules/http-cloud-function/variables.tf` line 57 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073#-Obi5AC-E_naTSVpYGOr:-OcjdGFqC-xx5SfEfUAF:bp73phl) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/96e6d9bbdcf4a8fe451ed402c3beb234811bfbda/src/main/terraform/gcloud/modules/http-cloud-function/variables.tf#L57)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> I couldn't find any trigger on the documentation that supports the triggering based on the file content

Doesn't `source_md5hash`[^1] do this?

[^1]: https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/storage_bucket_object#source_md5hash-1
</blockquote></details>

so that `source_md5hash` is available from version 6.41.0

I tried upgrading the terraform provider to re-generate `.terraform.lock.hcl` by using `git init`

but I have an error: `Error: storage.NewClient() failed: dialing: google: could not find default credentials.`

I cannot run that unless I download the json key file from google console which I believe you are opposed to.
For internal policy I cannot run `gcloud auth application-default login` to get those credentials, so I'm stuck here :)

are there other ways to regenerate that?
Otherwise can you upgrade it for me?


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: 1 of 8 files reviewed, 1 unresolved discussion (waiting on @Marco-Premier)

___
*[`src/main/terraform/gcloud/modules/http-cloud-function/variables.tf` line 57 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073#-Obi5AC-E_naTSVpYGOr:-Od9h3Mq6OqO8psvHJC4:b-ykpst0) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/96e6d9bbdcf4a8fe451ed402c3beb234811bfbda/src/main/terraform/gcloud/modules/http-cloud-function/variables.tf#L57)):*
> For internal policy I cannot run `gcloud auth application-default login` to get those credentials, so I'm stuck here :)

That's unfortunate, as that is the way you'd have to do it. Is this some Meta internal policy?

Regardless, the Terraform provider upgrade should be its own separate PR. That's not something that can be rolled back, so we need to be careful about it. I filed #3122. If you can't find a way to run `gcloud auth application-default login`, then you can assign that to me.


<!-- Sent from Reviewable.io -->

stevenwarejones reviewed (approved):
> @stevenwarejones reviewed 2 of 5 files at r1, 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: all files reviewed, 1 unresolved discussion (waiting on @SanjayVas)

___
*[`src/main/terraform/gcloud/modules/http-cloud-function/variables.tf` line 57 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073#-Obi5AC-E_naTSVpYGOr:-OdT_O4F1BibvegdcUoh:b-yvpzb5) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/96e6d9bbdcf4a8fe451ed402c3beb234811bfbda/src/main/terraform/gcloud/modules/http-cloud-function/variables.tf#L57)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> For internal policy I cannot run `gcloud auth application-default login` to get those credentials, so I'm stuck here :)

That's unfortunate, as that is the way you'd have to do it. Is this some Meta internal policy?

Regardless, the Terraform provider upgrade should be its own separate PR. That's not something that can be rolled back, so we need to be careful about it. I filed #3122. If you can't find a way to run `gcloud auth application-default login`, then you can assign that to me.
</blockquote></details>

yes - this is a meta restriction. can you help @SanjayVas  - I assigned that issue to you


<!-- Sent from Reviewable.io -->

stevenwarejones committed:
test: Apply PR #3073 changes for testing

Cherry-pick Marco's triggers_replace expansion onto current main
to test that config/env/secret changes trigger Cloud Function
redeployment.

Issue: world-federation-of-advertisers/cross-media-measurement#3073

stevenwarejones reviewed (commented):
> @stevenwarejones+AGNT made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: 5 of 8 files reviewed, 1 unresolved discussion (waiting on SanjayVas and stevenwarejones).

___
*[`src/main/terraform/gcloud/modules/http-cloud-function/variables.tf` line 57 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073#-Obi5AC-E_naTSVpYGOr:-OuwWogQ49s6NqLFqgGs:b-762jjz) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/96e6d9bbdcf4a8fe451ed402c3beb234811bfbda/src/main/terraform/gcloud/modules/http-cloud-function/variables.tf#L57)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

yes - this is a meta restriction. can you help @SanjayVas  - I assigned that issue to you
</blockquote></details>

Tested on Halo dev from branch `stevenwarejones_test-pr3073`. The `source_md5hash` + `uploaded_config_generation` approach works correctly.

Three `Update CMMS` runs with `apply=true` on dev:

1. **No config change** — no functions redeployed. Correct.
2. **Added comment to `DATA_WATCHER_CONFIG_CONTENT`** — `upload_data_watcher_config` detected the change via `source_md5hash`, new GCS generation propagated through `uploaded_config_generation`, all functions redeployed (expected on first run with new trigger values in state).
3. **Restored original config** — only `data-watcher` redeployed (15:22:07). `data-watcher-delete` was NOT redeployed (stayed at 15:15:48). Selective trigger working correctly.

Config has been restored to its original value. Dev is clean.


<!-- Sent from Reviewable.io -->

stevenwarejones reviewed (approved):
> @stevenwarejones reviewed 8 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: all files reviewed, 1 unresolved discussion (waiting on SanjayVas).



<!-- Sent from Reviewable.io -->

SanjayVas reviewed (approved):
> @SanjayVas reviewed 8 files and all commit messages, and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/3073)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on Marco-Premier).



<!-- Sent from Reviewable.io -->

stevenwarejones committed:
Merge branch 'main' into marcopremier/update_cloud_function_deployment
